### PR TITLE
overhaul authentication flow

### LINKED ISSUE #1318: ch-image:  Unable to pull from nVidia's NGC registry server
While attempting to download one of nVidia's "official" HPL-based HPC benchmarking containers, I found that `ch-image` was unable to correctly authenticate with the registry API.  From what I can tell, the NGC registry server is either using an entirely different authentication method than what `ch-image` supports or wanting to utilize a different auth mechanism without indicating that in a way `ch-image` currently recognizes.

The initial attempt:
```bash
> ch-image -vvv pull nvcr.io/nvidia/hpc-benchmarks:20.10-hpl .  
verbose level: 3
host architecture from uname: x86_64
host architecture for registry: amd64
found storage dir v2: /var/tmp/mej.ch
image_ref
  ir_hostport   nvcr.io
  ir_path       nvidia
  ir_name       hpc-benchmarks
  ir_tag        20.10-hpl

pulling image:    nvcr.io/nvidia/hpc-benchmarks:20.10-hpl
requesting arch:  amd64
destination:      .
downloading image: nvcr.io/nvidia/hpc-benchmarks:20.10-hpl
initializing session
auth: no authorization
GET: https://nvcr.io:443/v2/nvidia/hpc-benchmarks/manifests/20.10-hpl
response status: 401
HTTP 401 unauthorized
authorizing
WWW-Authenticate raw: Bearer realm="https://nvcr.io/proxy_auth",scope="repository:nvidia/hpc-benchmarks:pull,push"
WWW-Authenticate parsed: Bearer {'realm': 'https://nvcr.io/proxy_auth', 'scope': 'repository:nvidia/hpc-benchmarks:pull,push'}
authenticating using Bearer
requesting anonymous auth token
GET: https://nvcr.io/proxy_auth
response status: 200
received auth token: eyJraWQiOiJFUkNPOklCWFY6TjY2SDpO
retrying with auth: Bearer eyJraWQiOiJFUkNPOklCWFY6TjY2SDpO
GET: https://nvcr.io:443/v2/nvidia/hpc-benchmarks/manifests/20.10-hpl
response status: 401
b'<html>\r\n<head><title>401 Authorization Required</title></head>\r\n<body bgcolor="white">\r\n<center><h1>401 Authorization Required</h1></center>\r\n<hr><center>nginx/1.14.2</center>\r\n</body>\r\n</html>\r\n'
error: not in registry: nvcr.io:443/nvidia/hpc-benchmarks:20.10-hpl
```

I was expecting `ch-image` to prompt for the username and password for authentication, but it did not, even though its first `GET` attempt got a `401 unauthorized` response back.

Supplying the credentials in the environment as specified in the Charliecloud docs resulted in very similar behavior (and still failed):
```bash
>> (export CH_IMAGE_USERNAME='$oauthtoken' CH_IMAGE_PASSWORD='' ; read -esp 'Registry Server Password:  ' CH_IMAGE_PASSWORD && ch-image -vvv pull nvcr.io/nvidia/h
pc-benchmarks:20.10-hpl .)                                                                 
verbose level: 3Password:                                                                 
host architecture from uname: x86_64
host architecture for registry: amd64
found storage dir v2: /var/tmp/mej.ch
image_ref                                                                                                                                                                           
  ir_hostport   nvcr.io                                                                                                                                                             
  ir_path       nvidia     
  ir_name       hpc-benchmarks 
  ir_tag        20.10-hpl      
                                             
pulling image:    nvcr.io/nvidia/hpc-benchmarks:20.10-hpl
requesting arch:  amd64                                                                   
destination:      .                                                                       
downloading image: nvcr.io/nvidia/hpc-benchmarks:20.10-hpl
initializing session                                                                                                                                                                
auth: no authorization          
GET: https://nvcr.io:443/v2/nvidia/hpc-benchmarks/manifests/20.10-hpl
response status: 401
HTTP 401 unauthorized                                                                                                                                                               
authorizing          
WWW-Authenticate raw: Bearer realm="https://nvcr.io/proxy_auth",scope="repository:nvidia/hpc-benchmarks:pull,push"
WWW-Authenticate parsed: Bearer {'realm': 'https://nvcr.io/proxy_auth', 'scope': 'repository:nvidia/hpc-benchmarks:pull,push'}
authenticating using Bearer
requesting anonymous auth token
GET: https://nvcr.io/proxy_auth
response status: 200
received auth token: eyJraWQiOiJFUkNPOklCWFY6TjY2SDpO
retrying with auth: Bearer eyJraWQiOiJFUkNPOklCWFY6TjY2SDpO
GET: https://nvcr.io:443/v2/nvidia/hpc-benchmarks/manifests/20.10-hpl
response status: 401
b'<html>\r\n<head><title>401 Authorization Required</title></head>\r\n<body bgcolor="white">\r\n<center><h1>401 Authorization Required</h1></center>\r\n<hr><center>nginx/1.14.2</ce
nter>\r\n</body>\r\n</html>\r\n'
error: not in registry: nvcr.io:443/nvidia/hpc-benchmarks:20.10-hpl
```

So `docker login && docker pull nvcr.io/nvidia/hpc-benchmarks:20.10-hpl` functions correctly, but `ch-image pull nvcr.io/nvidia/hpc-benchmarks:20.10-hpl .` fails.

### PR DESCRIPTION
**Pull request by reidpr**
Closes #1318? See also PR #1324.

### CODE REVIEW

**Comment by reidpr** [comment]
OK, here's a different proposed behavior. @adrecord, what's your take?

**Principles**

1. Charliecloud does not store credentials like Docker and Podman, and I’m disinclined to start storing any configuration, let alone secrets. If we accept this constraint, we can’t do the permanent `docker login` model. But, it’s possible to simulate it with the environment variables.

2. Usability: We want to avoid prompting for credentials if the authenticated request would also not succeed. This is impossible, because we don’t know if it would succeed until we authenticate and try it. However, how close can we come?

**Authentication models**

I saw several authentication models:

*Never*: Only try anonymous access; never try to authenticate.

*Sloppy Lazy*: Try anonymous first. If that fails, and we *suspect* authenticating might help, try to authenticate. If authentication successful, re-try the request that failed.

* This is the current behavior, but it fails on nVidia NGC (issue #1318) and possibly other registries. In principle we could improve the heuristic, but that seems hard.

*Lazy*: Try anonymous first. If that fails, and there is *any chance* authenticated might succeed, try to authenticate.

* AFAICT, this is the behavior proposed in PR #1324. I’m concerned it would yield excessive prompts for a password, e.g. if you `ch-image pull debian:bullsear` (instead of the correct `bullseye`), it will prompt for a password, try again, and still fail.

*Eager*: Try authenticating before making any requests. If this succeeds, then make all the substantive requests in authenticated mode; otherwise, make them anonymously.

* Problem: If the user just typoes the username or password, we can’t tell the difference between that and no account existing. Thus, there’s no retry of username/password within Charliecloud, and error messages might not be clear. The user would have to re-run ch-image.

*Always*: Try authenticating before making any requests. If this fails, error out. Otherwise, make all requests authenticated.

**Options**

1. Keep the current behavior (sloppy lazy). This seems bad because it doesn't work on all registries.

2. Provide comprehensive options, e.g. `--auth={never,lazy,eager,always,auto}`

3. Provide `--auth` and/or `--no-auth`. There are some sub-options:

   * `--auth` ⇒ eager vs. `--auth` ⇒ always
   * `--no-auth` ⇒ lazy vs. `--no-auth` ⇒ never
   * What is the default behavior if neither are specified?
   * Should the default change if the environment variables are specified?
   * Should we also have an environment variable e.g. `$CH_IMAGE_AUTH` to set the default?

Currently, this PR provides `--auth` to select *always* mode, a default mode of never, `--no-auth` not supported, and `$CH_IMAGE_AUTH` to set a persistent mode of *always*.

File: lib/charliecloud.py [hunk 1]
```diff
@@ -1239,42 +1243,235 @@ def write(self, data):
...
+   #                 classes that may not have been defined when the module is
+   #                 created, e.g. classes later in the file, or some can
+   #                 escalate to themselves.
+   #
+   #   reg_auth .... True if appropriate for authenticated mode, False if
+   #                 anonymous (i.e., --auth or not, respectively). Everything
+   #                 must be one or the other.
+   #
+   #   name ........ Auth type string (from WWW-Authenticate header) that this
+   #                 class matches.
```
Comment (j-ogas) [inline lib/charliecloud.py:1271]:
```suggestion
   #   scheme ...... Auth scheme string (from WWW-Authenticate header) that
   #                 this class matches.
```

File: lib/charliecloud.py [hunk 2]
```diff
@@ -1239,42 +1243,235 @@ def write(self, data):
...
+      (auth_type, auth_d) = auth_h.split(maxsplit=1)
+      auth_d = urllib.request.parse_keqv_list(
+                  urllib.request.parse_http_list(auth_d))
+      VERBOSE("WWW-Authenticate parsed: %s %s" % (auth_type, auth_d))
+      # Is escalation possible in principle?
+      if (len(self.escalators) == 0):
+         FATAL("no further authentication possible, giving up")
+      # Try to escalate.
+      for class_ in self.escalators:
+         if (class_.name == auth_type):
```
Comment (j-ogas) [inline lib/charliecloud.py:1313]:
```suggestion
      (auth_scheme, auth_d) = auth_h.split(maxsplit=1)
      auth_d = urllib.request.parse_keqv_list(
                  urllib.request.parse_http_list(auth_d))
      VERBOSE("WWW-Authenticate parsed: %s %s" % (auth_scheme, auth_d))
      # Is escalation possible in principle?
      if (len(self.escalators) == 0):
         FATAL("no further authentication possible, giving up")
      # Try to escalate.
      for class_ in self.escalators:
         if (class_.scheme == auth_scheme):
```

File: lib/charliecloud.py [hunk 3]
```diff
@@ -1239,42 +1243,235 @@ def write(self, data):
...
+                  VERBOSE("authentication did not escalate; trying next")
+               else:
+                  return auth  # success!
+      VERBOSE("no authentication left to try")
+      return None
+
+
+class Registry_Auth_Basic(Registry_Auth):
+   anon_p = False
+   name = "Basic"
```
Comment (j-ogas) [inline lib/charliecloud.py:1331]:
```suggestion
   scheme = "Basic"
```

File: lib/charliecloud.py [hunk 4]
```diff
@@ -1239,42 +1243,235 @@ def write(self, data):
...
+      (username, password) = reg.creds.get()
+      i = class_()
+      i.basic = requests.auth.HTTPBasicAuth(username, password)
+      return i
+
+
+class Registry_Auth_Bearer_IDed(Registry_Auth):
+   # https://stackoverflow.com/a/58055668
+   anon_p = False
+   name = "Bearer"
```
Comment (j-ogas) [inline lib/charliecloud.py:1363]:
```suggestion
   scheme = "Bearer"
```

File: lib/charliecloud.py [hunk 5]
```diff
@@ -1239,42 +1243,235 @@ def write(self, data):
...
+   def token_auth(class_, creds):
+      """Return a requests.auth.AuthBase object used to authenticate the token
+         request."""
+      (username, password) = creds.get()
+      return requests.auth.HTTPBasicAuth(username, password)
+
+
+class Registry_Auth_Bearer_Anon(Registry_Auth_Bearer_IDed):
+   anon_p = True
+   name = "Bearer"
```
Comment (j-ogas) [inline lib/charliecloud.py:1426]:
```suggestion
   scheme = "Bearer"
```

File: lib/charliecloud.py [hunk 6]
```diff
@@ -1239,42 +1243,235 @@ def write(self, data):
...
+   @classmethod
+   def token_auth(class_, creds):
+      # The way to get an anonymous Bearer token is to give no Basic auth
+      # header in the token request.
+      return None
+
+
+class Registry_Auth_None(Registry_Auth):
+   anon_p = True
+   name = None
```
Comment (j-ogas) [inline lib/charliecloud.py:1444]:
```suggestion
   scheme = None
```

File: lib/charliecloud.py [hunk 7]
```diff
@@ -1239,42 +1243,235 @@ def write(self, data):
...
+   #reg_auth =   # starting authentication in both modes
```
Comment (j-ogas) [inline lib/charliecloud.py:1445]: Do we need this comment?

File: lib/charliecloud.py [hunk 8]
```diff
@@ -1495,18 +1640,27 @@ def request(self, method, url, statuses={200}, out=None, **kwargs):
...
+         if (res.status_code != 401):
+            break
+         else:
+            VERBOSE("HTTP 401 unauthorized")
+            if (self.escalate(res)):   # success
+               VERBOSE("retrying with auth: %s" % self.auth)
+            elif (401 in statuses):    # caller can deal with it
+               break
+            else:
+               FATAL("authorization failure; don't know what to do")
```
Comment (j-ogas) [inline lib/charliecloud.py:1662]:
```suggestion
               FATAL("authorization failure; all auth methods exhausted")
```

**Review by j-ogas — CHANGES_REQUESTED** [review]
See nitpicky suggestions.

**Review by j-ogas — COMMENTED** [review]
Living dangerously I see, e.g., skipping CI on my suggestions?

One thing I don't see here are updates to `/test/build/40_pull.bats`, e.g., pull from previously troublesome registries like Nvidia. Do we care?

If we don't care to add a registry test and we are comfortable merging the UI suggestions without CI, then I give it my blessing.

**Review by j-ogas — APPROVED** [review]
LGTM.

**Pull request merged**